DOC: Update dev workflow example (#10849)
astropy/astropy: docs/development/workflow/git_edit_workflow_examples.rst

A few steps in this tutorial take only a single command. They are broken out
separately to outline the process in words as well as code.

Inform your local copy of ``astropy`` about the latest changes in the development
version with::

    git fetch astropy --tags

Set up an isolated workspace
============================

+ Make a new `git`_ branch for fixing this issue, check it out, and let my
  GitHub account know about this branch::

    git checkout astropy/master -b fix-1761 # branch based on latest from GitHub
                                            # and switch to this branch

>     git push your-user-name fix-1761        # tell my GitHub acct about it

Reviewer: I'm confused why this is here (and in the original). If you do it later, when the work is ready, then github will give you a nice web link to copy & paste to do the PR.

Also, if you keep it here, I would keep the `--set-upstream`...

Reviewer 2: for what it's worth, I never manually say `--set-upstream`. Otherwise I agree that it might be better to push only once some work is done to get the github pop-up for opening the PR (maybe it also works when pushing for an existing branch?, I've never tried).

Author: I never use `--set-upstream`. In fact, I think a plain old `git push` is dangerous (personal opinion) and prefer to be explicit when I do the push. If there is a strong preference to add it back, I will.

Sure, I can also delay `git push` till the end. I wasn't questioning why this tutorial was pushing so often.

Reviewer 2: oh, `git push`, just as is, is only asking for the troubles, I don't think should advocate for it anywhere.

Reviewer: I do use `-u` (I find I'm more likely to push to the wrong branch if I have to enter a name, even with auto-completion...), but for the dev workflow example, I agree it is better not to - keep things as simple as possible. And really it is only useful if one keeps on pushing. So, if you are up for it, I would remove the git push until after the user has done the testing. After all, the whole point of git is to enable asynchronous development - there is no reason to keep github in the loop.

Reviewer 2: @mhvk - I use gitprompt that decorates my bash prompt (there are way more fancy solutions, too, I just never got to the point of switching), so I always see which branch I'm at. It gives a very convenient double check for names, as well as branch status compared to upstream/origin. And I just got into the habit of double checking with a git gui whenever I need to force push to a branch (in those cases I usually just see the hash as normally I don't bother to create a branch when working with other people's forks). 
Overall I find this a highly efficient, but also a quite safe setup.

Author: I am not sure if we want to get in the weeds on setting up custom git prompt etc. But I did remove all the extraneous `git push` commands.

Reviewer 2: Oh, I'm not saying we should do anything about git prompt, it's on the same level of arbitrary personal preferences as the choice of editor. It was only a pointer to Marten's point that helps to avoid pushing to the wrong branch, in case he doesn't use it or something similar.

+ Make a Python environment just for this fix and switch to that environment.
  The example below shows the necessary steps in the Miniconda/Anaconda Python
  distribution::

>     conda create -n apy-1761 python=3.9 # replace 3.9 with desired version

Reviewer: Since we now use `tox` for testing, which sets up its own environment anyway, I think we should remove this as unnecessary complication...

Author: Given #10851 , I am hesitant to introduce `tox` here. Can we do the `tox` stuff as a follow-up PR and let someone who really likes `tox` do it? (I don't use `tox`.)

Reviewer: OK, but doesn't one somewhere actually need to do `setup.py develop`, so that the C stuff is properly compiled? Or are we relying on auto-compilation here?

Author: `pip install -e .` has always worked for me. It also compiles the C stuff. I never do `setup.py develop` myself.

    conda activate apy-1761

  If you are using a different distribution, see :ref:`virtual_envs` for
  instructions for creating and activating a new environment.

+ Install our branch in this environment with::

>     pip install -e .

Reviewer 3: Later in the tutorial we ask people to run pytest, we should have them install it here. You could argue for a `[all,test]` here if you want to be able to run the full suite.
```suggestion
    pip install -e .[test]
```

Reviewer 3: @bsipocz This is my current leading theory for the issues the new astroquery contributor ran into fyi.

Reviewer 3: @mhvk's suggestion below to turn the pytest into a call to tox is way better. Although arguably we should have them install the test extra anyway.

Do you really have to set up a separate Python environment for each fix? No,
but you definitely want to have a Python environment for your work on code
contributions. Making new environments is fast, does not take much space, and
provide a way to keep your work organized.

If installation fails, try to upgrade ``pip`` using ``pip install pip -U``
command. It is also a good practice to keep your ``conda`` up-to-date by
running ``conda update conda -n base`` when prompted to do so; maybe ``git`` too.

> 

Reviewer: So I'd recommend removing the above until here....

command. It is also a good practice to keep your ``conda`` up-to-date by
running ``conda update conda -n base`` when prompted to do so; maybe ``git`` too.

Test first, please
==================

It would be hard to overstate the importance of testing to Astropy. Tests are
It would be hard to overstate the importance of testing in Astropy. Tests are
what gives you confidence that new code does what it should and that it
doesn't break old code.
does not break old code.

You should at least run the relevant tests before you make any changes to make
sure that your python environment is set up properly.
sure that your Python environment is set up properly.

The first challenge is figuring out where to look for relevant tests. `Issue
1761`_ is a problem in the `~astropy.coordinates` package, so the tests for
it are in ``astropy/coordinates/tests``. The rest of Astropy has a similar
layout, described at :ref:`testing-guidelines`.
it are in ``astropy/coordinates/tests``. The rest of ``astropy`` has a similar
layout, as described at :ref:`testing-guidelines`.

Change to that directory and run the current tests with::

>     cd astropy/coordinates/test

Reviewer: Ouch, this is quite bad!!!  Doesn't even install things beforehand.  I'd say this should be `tox -e test -- -P coordinates`...

Or, after all, suggest setting up with `tox --develop`.....

Reviewer 3: It should work to run pytest directly, but yeah for a first time contributor I would suggest tox, no develop etc.

>     cd astropy/coordinates/tests
>     pytest

Reviewer 3: This should really be:
```suggestion
    pytest astropy/coordinates/tests
```
or via tox.

Author: I removed instructions to keep changing directories, so they all run from top-level now.

> A subpackage organize its tests into multiple test modules; e.g.::

Author: Note to self: Fold this into the next commits after reviews come in.
```suggestion
A subpackage organizes its tests into multiple test modules; e.g.::
```

new test in ``test_arrays.py``. As with all of the steps, when in doubt,
please ask on the `astropy-dev mailing list`_.

The goal at this point may be a little counter-intuitive: write a test that we
know will fail with the current code. This test allows ``astropy`` to check,
in an automated way, whether our fix actually works and to prevent
regression (i.e., make sure future changes to code do not break our fix).

Looking over the existing code in ``test_arrays.py``, each test is a function

> whose name starts with ``test_``. An appropriate place to add the test is

Author: ```suggestion
with a name that starts with ``test_``. An appropriate place to add the test is
```
